### DISCUSSION TITLE
Node meta labels are missing when converting existing CAdvisor configuration to service monitor

### DISCUSSION THREAD

**Discussion by anoop2503** [prometheus-operator/prometheus-operator#5749] [Q&A]
Hi,

As part of moving to Prometheus operator, I am trying to migrate the existing CAdvisor job configuration from current Prometheus to use Service Monitor in Prometheus Operator. The existing configuration is using 'node' role in kubernetes_sd_config, since we need node related meta labels for monitoring. But, when configuring Service Monitor with endpoints role, these node related meta labels are missing in the metrics. Is there anyway to use node role with service monitor or fix this issue some other way?

Thanks

**Comment by ArthurSens** [comment]
ServiceMonitors use the `endpoints` role when using Kubernetes service discovery. There are 2 approaches here:

* You can use [ScrapeConfig](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com/v1alpha1.ScrapeConfig), which is possible to configure the `node` role with Kubernetes service-discovery.
* You can use ServiceMonitor and scrape a service that exposes the cAdvisor metrics endpoint. There are examples in our subproject, [kube-prometheus](https://github.com/prometheus-operator/kube-prometheus/blob/main/manifests/kubernetesControlPlane-serviceMonitorKubelet.yaml) :)

**Reply by anoop2503** [reply]
Thanks @ArthurSens for checking this.

I am already following the second option what you mentioned. However, since we are mentioning "endpoints" role in the service monitor, the node related meta labels (eg: nodegroup, __meta_kubernetes_node_*) not present in the timeseries. (Ref: https://github.com/prometheus-operator/kube-prometheus/blob/main/manifests/kubernetesControlPlane-serviceMonitorKubelet.yaml#L10)

**Reply by ArthurSens** [reply]
Oh yes, sorry. Maybe using PodMonitors could be an option, which uses the `pod` role, if the pod has node related labels.

Otherwise I'd go with ScrapeConfig, which has the `node` role.

**Reply by anoop2503** [reply]
Sure, thanks @ArthurSens. 
Is there any sample configuration available online for ScrapeConfig?

**Reply by ArthurSens** [reply]
Work in progress here 😅 
https://github.com/prometheus-operator/prometheus-operator/pull/5659

**Reply by anoop2503** [reply]
Great, thank you @ArthurSens.
